3Dmol/glviewer.js

```diff
@@ -3360,46 +3358,55 @@ $3Dmol.GLViewer = (function() {
-                            surfobj.done = true;
-                            if(surfacecallback && typeof(surfacecallback) == "function") {
-                                surfacecallback(surfid);
+                    var rfunction = function(event_data) {
+                        for (var j=0;j<event_data.length;j++) {
+                            var VandFs = $3Dmol.splitMesh({vertexArr:event_data[j].vertices,
+                                                           faceArr:event_data[j].faces});
+                            for(var i=0,vl=VandFs.length;i<vl;i++){
+                                var VandF={vertices:VandFs[i].vertexArr,
+                                           faces:VandFs[i].faceArr};
+                                var mesh = generateSurfaceMesh(atomlist, VandF, mat);
+                                $3Dmol.mergeGeos(surfobj.geo, mesh);
                             }
+                            _viewer.render();
                         }
                     };
 
                     var efunction = function(event) {
                         console.log(event.message + " (" + event.filename + ":" + event.lineno + ")");
                     };
 
-                    for (i = 0; i < extents.length; i++) {
-                        var worker = workers[i % workers.length];
-                        worker.onmessage = rfunction;
-
-                        worker.onerror = efunction;
-
-                        worker.postMessage({
-                            'type' : type,
-                            'expandedExtent' : extents[i].extent,
-                            'extendedAtoms' : extents[i].atoms,
-                            'atomsToShow' : extents[i].toshow
+                    var startWorker = function(i) {
+                        return new Promise(function(resolve, reject) {
+                            var worker = workers[i % workers.length];
+                            worker.onmessage = function(event) {
+                                resolve(event.data);
+                            }
+                            worker.onerror = function(event) {
+                                reject(event);
+                            }
+                            worker.postMessage({
+                                'type' : type,
+                                'expandedExtent' : extents[i].extent,
+                                'extendedAtoms' : extents[i].atoms,
+                                'atomsToShow' : extents[i].toshow
+                            });
                         });
                     }
+                    var promises = []
+                    for (i = 0; i < extents.length; i++) {
+                        promises.push(startWorker(i));
+                    }
+
+                    return Promise.all(promises)
+                    .then(function(event_data) {
+                        rfunction(event_data);
+                        surfobj.done = true;
```

Comment:
Since, I am using Promise.all, I am not sure how I can change here to return only the surfid, instead of an array of surfid. So, I changed the promise at the end to resolve using surfid.